File: python/Terminal.py
import serial
import json

class Terminal:

    terminal_type = "v400c-plus"
    terminal = serial.Serial()

    def __init__(self, terminal_type='v400c-plus', baudrate=9600, bytesize=8, stopbit=1, xonxoff=0, rtscts=0, port=4, timeout=0):
        acceptable_terminal_types = ['v200c', 'v400c', 'v400c-plus']

        if terminal_type not in acceptable_terminal_types:
            raise Exception("invalid terminal type [" + terminal_type + "]")

        self.terminal_type = terminal_type
        self.terminal.baudrate = baudrate
        self.terminal.bytesize = bytesize
        self.terminal.stopbit = stopbit
        self.terminal.xonxoff = xonxoff
        self.terminal.rtscts = rtscts
        self.terminal.timeout = timeout
        self.terminal.port = port
        self.terminal.parity = serial.PARITY_NONE

    def process(self, action, amount, options=None):
        action = action.upper()
        acceptable_action_types = ["SALE", "REFUND", "VOID"]
        request_json = {"Amount": amount}

        if action not in acceptable_action_types:
            raise Exception("invalid terminal action [" + action + "]")

        if action == "VOID" and "PaymentID" not in options:
            raise Exception(action + " requires a PaymentID")

        request_json["Operator"] = action

        for field in ["Tip", "Invoice", "Clerk", "PaymentItem", "PaymentID"]:
            if field in options:
                request_json[field] = options[field]

        # Send request to terminal
        self.terminal.open()

        response = ""
        if self.terminal.is_open:
            self.terminal.flushInput()
            self.terminal.flushOutput()
            self.terminal.write(json.dumps(request_json))
            while True:
                response = self.terminal.readline().decode('ascii')
                break;
            self.terminal.close()

        return response

    def sale(self, amount, options=None):
        return self.process("SALE", amount, options)

    def refund(self, amount, options=None):
        return self.process("REFUND", amount, options)

    def void(self, amount, options=None):
        return self.process("VOID", amount, options)

File: c++/terminal.cpp



#include "terminal.h"

  namespace IntelliPayTerminal {

      inline bool INTELLIPAY_DLL init (string port, string terminal_type = "v400c-plus", int baud_rate = 9600, serial::bytesize_t byte_size = serial::eightbits, serial::stopbits_t stop_bit = serial::stopbits_one, serial::flowcontrol_t xonxoff = serial::flowcontrol_none, bool rtscts = false, int timeout = 0) {
        try {

            intellipay_terminal_type = terminal_type;
            intellipay_serial_port = port;

            // let's set up the connection to the terminal device.
            stty.setPort(intellipay_serial_port);
            stty.setBaudrate(baud_rate);
            stty.setStopbits(stop_bit);
            stty.setBytesize(byte_size);
            stty.setFlowcontrol(xonxoff);
            stty.setRTS(rtscts);
            stty.open();
            if (!stty.isOpen()) {
                return false;
            }
        } catch (const exception e) {
            return false;
        }
        return true;
    }

    inline json INTELLIPAY_DLL process(string action, float amount, json options = json({})) {
        boost::algorithm::to_upper(action);
        if (action != "SALE" && action != "REFUND" && action != "VOID") {
            throw "invalid action.";
        }

        json obj;
        obj["Operation"] = action;

        string search_fields[5] = {"Tip", "Invoice", "Clerk", "PaymentItem", "PaymentID"};

        for (int x = 0; x < search_fields->length(); x++) {
            if (options.contains(search_fields[x])) {
                obj[search_fields[x]] = options[search_fields[x]];
            }
        }

        json response;
        response["status"] = "failed";

        if (!stty.isOpen()) {
            stty.open();

            if (!stty.isOpen())
                return response;
        }

        stty.write(obj.dump());

        string read_buffer;
        stty.readline(read_buffer, 256, "\n");
        if (!read_buffer.length()) {
            response["message"] = "Error reading from terminal device.";
        }

        json terminal_response = json::parse(read_buffer);

        stty.close();
        if (terminal_response.contains("Response")) {
            if (terminal_response["Response"].contains("PaymentID")) {
                response["payment-id"] = terminal_response["Response"]["PaymentID"];
            }

            if (terminal_response["Response"].contains("ResponseCode") && terminal_response["Response"]["ResponseCode"] == "00") {
                response["status"] = "processed";
            }

            if (terminal_response["Response"].contains("Operation") &&
                (terminal_response["Response"]["Operation"] == "SALE" || terminal_response["Response"]["Operation"] == "REFUND") ){
                response["card-information"] = json({});
                response["card-information"]["brand"] = terminal_response["Response"]["Card"]["Brand"];
                response["card-information"]["last4"] = terminal_response["Response"]["Card"]["Last4"];
                response["card-information"]["expiration"] = terminal_response["Response"]["Card"]["ExpDate"];
            }
        }
        return response;
    }

    inline json INTELLIPAY_DLL sale(float amount, json options = json({})) {
        return process("sale", amount, options);
    }

    inline json INTELLIPAY_DLL refund(float amount, json options = json({})){
        return process("refund", amount, options);
    }
}